Missing ')'

Hello again, this time with a real bug, I think.

```
/virtual/main.c:66:83: error: expected ')'
        bpf_dins_pkt(skb, (u64)ip+12, 0, 32, ibpf_dext_pkt(skb, (u64)ip+16, 0, 32);
                                                                                  ^
/virtual/main.c:66:21: note: to match this '('
        bpf_dins_pkt(skb, (u64)ip+12, 0, 32, ibpf_dext_pkt(skb, (u64)ip+16, 0, 32);
```

Here is a sample program that is causing the error.

```
#include <uapi/linux/bpf.h>
#include <uapi/linux/if_ether.h>
#include <uapi/linux/if_packet.h>
#include <uapi/linux/ip.h>
#include <uapi/linux/in.h>
#include <uapi/linux/udp.h>
#include <bcc/proto.h>

#define ETH_LEN 14

struct dns_hdr_t
{
    uint16_t id;
    uint16_t flags;
    uint16_t qdcount;
    uint16_t ancount;
    uint16_t nscount;
    uint16_t arcount;
} BPF_PACKET_HEADER;


struct dns_query_flags_t
{
  uint16_t qtype;
  uint16_t qclass;
} BPF_PACKET_HEADER;

struct dns_char_t
{
    char c;
} BPF_PACKET_HEADER;

struct Key {
  unsigned char p[32];
};

struct Leaf {
  // Not really needed in this example
  unsigned char p[4];
};

BPF_TABLE("hash", struct Key, struct Leaf, cache, 128);

int dns_test(struct __sk_buff *skb)
{
  u8 *cursor = 0;
  struct Key key = {};
  struct ethernet_t *ethernet = cursor_advance(cursor, sizeof(*ethernet));
  if(ethernet->type == ETH_P_IP) {

    struct ip_t *ip = cursor_advance(cursor, sizeof(*ip));
    u16 hlen_bytes = ip->hlen << 2;
    if(ip->nextp == IPPROTO_UDP) {

      struct udp_t *udp = cursor_advance(cursor, sizeof(*udp));
      if(udp->dport == 53){
        u8 *sentinel = cursor + udp->length - sizeof(*udp) - 4;

        struct dns_hdr_t *dns_hdr = cursor_advance(cursor, sizeof(*dns_hdr));
        if((dns_hdr->flags >>15) != 0) {
          return -1;
        }
         // Error is happening here.
          ip->src = ip->dst;
          return 0;
      }
    }
  }

  return -1;
}
```

Hi @drzaeus77, there are also some issues when defines are used:

```
#include <bcc/proto.h>

#define ETHERNET_BROADCAST 0xffffffffffff

int http_filter(struct __sk_buff *skb) {
    u8 *cursor = 0;
    struct ethernet_t *ethernet = cursor_advance(cursor, sizeof(*ethernet));
    ethernet->src = ETHERNET_BROADCAST;
    return -1;
}
```
fails to compile with the following error:
```
/virtual/main.c:14:44: error: expected expression
        bpf_dins_pkt(skb, (u64)ethernet+6, 0, 48, )     ethernet->src = ETHERNET_BROADCAST;
```

A workaround is to enclose the define in parentheses:
``` 
ethernet->src = (ETHERNET_BROADCAST);
```
PD: please ignore any possible verifier error on this example, I created it only to illustrate this bcc issue.

--- a/src/cc/frontends/clang/b_frontend_action.cc
+++ b/src/cc/frontends/clang/b_frontend_action.cc
@@ -521,7 +521,6 @@ bool BTypeVisitor::VisitBinaryOperator(BinaryOperator *E) {
   if (!E->isAssignmentOp())
     return true;
   Expr *LHS = E->getLHS()->IgnoreImplicit();
-  Expr *RHS = E->getRHS()->IgnoreImplicit();
   if (MemberExpr *Memb = dyn_cast<MemberExpr>(LHS)) {
     if (DeclRefExpr *Base = dyn_cast<DeclRefExpr>(Memb->getBase()->IgnoreImplicit())) {
       if (DeprecatedAttr *A = Base->getDecl()->getAttr<DeprecatedAttr>()) {
@@ -534,10 +533,10 @@ bool BTypeVisitor::VisitBinaryOperator(BinaryOperator *E) {
             uint64_t ofs = C.getFieldOffset(F);
             uint64_t sz = F->isBitField() ? F->getBitWidthValue(C) : C.getTypeSize(F->getType());
             string base = rewriter_.getRewrittenText(expansionRange(Base->getSourceRange()));
-            string rhs = rewriter_.getRewrittenText(expansionRange(RHS->getSourceRange()));
             string text = "bpf_dins_pkt(" + fn_args_[0]->getName().str() + ", (u64)" + base + "+" + to_string(ofs >> 3)
-                + ", " + to_string(ofs & 0x7) + ", " + to_string(sz) + ", " + rhs + ")";
-            rewriter_.ReplaceText(expansionRange(E->getSourceRange()), text);
+                + ", " + to_string(ofs & 0x7) + ", " + to_string(sz) + ",";
+            rewriter_.ReplaceText(expansionRange(SourceRange(E->getLocStart(), E->getOperatorLoc())), text);
+            rewriter_.InsertTextAfterToken(E->getLocEnd(), ")");
           }
         }
       }

--- a/tests/python/test_clang.py
+++ b/tests/python/test_clang.py
@@ -392,5 +392,24 @@ int process(struct xdp_md *ctx) {
         t = b["act"]
         self.assertEquals(len(t), 32);
 
+    def test_bpf_dins_pkt_rewrite(self):
+        text = """
+#include <bcc/proto.h>
+int dns_test(struct __sk_buff *skb) {
+    u8 *cursor = 0;
+    struct ethernet_t *ethernet = cursor_advance(cursor, sizeof(*ethernet));
+    if(ethernet->type == ETH_P_IP) {
+        struct ip_t *ip = cursor_advance(cursor, sizeof(*ip));
+        ip->src = ip->dst;
+        return 0;
+    }
+    return -1;
+}
+        """
+        b = BPF(text=text)
+
 if __name__ == "__main__":
     main()
+
+
+

